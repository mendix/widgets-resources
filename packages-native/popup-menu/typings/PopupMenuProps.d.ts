/**
 * This file was generated from PopupMenu.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Team
 */
import { ActionValue } from "mendix";
import { ReactNode } from "react";

interface CommonProps<Style> {
    name: string;
    style: Style[];
}

export type RenderModeEnum = "basic" | "custom";

export type ItemTypeEnum = "item" | "divider";

export type StyleClassEnum = "default" | "primary" | "danger";

export interface BasicItemsType {
    itemType: ItemTypeEnum;
    caption?: string;
    action?: ActionValue;
    styleClass: StyleClassEnum;
}

export interface ComplexItemsType {
    content: ReactNode;
    action?: ActionValue;
}

export interface PopupMenuProps<Style> extends CommonProps<Style> {
    menuTriggerer?: ReactNode;
    renderMode: RenderModeEnum;
    basicItems: BasicItemsType[];
    complexItems: ComplexItemsType[];
}
