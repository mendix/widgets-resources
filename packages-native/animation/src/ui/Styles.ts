import { Style } from "@native-mobile-resources/util-widgets";
import { ViewStyle } from "react-native";

export interface AnimationStyle extends Style {
    container: ViewStyle;
}

export const defaultAnimationStyle: AnimationStyle = {
    container: {
        // All ViewStyle properties are allowed
    }
};
