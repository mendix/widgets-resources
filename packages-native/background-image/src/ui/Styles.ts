import { ViewStyle, ImageStyle } from "react-native";
import { Style } from "@native-mobile-resources/util-widgets";

export interface BackgroundImageStyle extends Style {
    container: ViewStyle;
    image: ImageStyle;
}

export const defaultBackgroundImageStyle: BackgroundImageStyle = {
    container: {},
    image: {}
};
