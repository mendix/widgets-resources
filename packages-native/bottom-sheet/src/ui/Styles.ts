import { Style } from "@native-mobile-resources/util-widgets";
import { ViewStyle } from "react-native";

export interface BottomDrawerStyle extends Style {
    container: ViewStyle;
}

export const defaultBottomDrawerStyle: BottomDrawerStyle = {
    container: {
        flex: 1
    }
};
