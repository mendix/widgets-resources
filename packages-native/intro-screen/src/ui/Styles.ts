import { Style } from "@native-mobile-resources/util-widgets";
import { I18nManager, Platform, StyleSheet, TextStyle, ViewStyle } from "react-native";
import absoluteFillObject = StyleSheet.absoluteFillObject;
import DeviceInfo from "react-native-device-info";

const isiPhoneModelWithNotch = (): boolean => {
    const model = DeviceInfo.getDeviceId();
    if (model.indexOf("iPhone") !== -1) {
        switch (model) {
            case "iPhone10,6": // iPhone X GSM
            case "iPhone11,2": // iPhone XS
            case "iPhone11,4": // iPhone XS Max
            case "iPhone11,6": // iPhone XS Max Global
            case "iPhone11,8": // Iphone XR
            case "iPhone12,1": // Iphone 11
            case "iPhone12,3": // Iphone 11 Pro
            case "iPhone12,5": // Iphone 11 Pro Max
                return true;
            default:
                return false;
        }
    }
    return false;
};

const isAndroidRTL = I18nManager.isRTL && Platform.OS === "android";
const isIphoneWithNotch = DeviceInfo.hasNotch() || (Platform.OS === "ios" && isiPhoneModelWithNotch());

const defaultButtonBetweenContainer: ViewStyle = {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    paddingVertical: 12,
    alignSelf: "stretch"
};

const defaultButtonBelowContainer: ViewStyle = {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    paddingVertical: 12
};

const defaultButtonCaption: TextStyle = {
    backgroundColor: "transparent",
    color: "black",
    fontSize: 18,
    alignSelf: "center"
};

export interface ButtonStyle {
    container: ViewStyle;
    caption: TextStyle;
    icon: {
        color: string;
        size?: number;
    };
}

export interface IntroScreenStyle extends Style {
    fullscreenContainer: ViewStyle;
    popupContainer: ViewStyle;
    paginationContainer: ViewStyle;
    paginationText: TextStyle;
    dotStyle: ViewStyle;
    activeDotStyle: ViewStyle;
    paginationAbove: {
        buttonsContainer: ViewStyle;
        buttonSkip: ButtonStyle;
        buttonDone: ButtonStyle;
        buttonPrevious: ButtonStyle;
        buttonNext: ButtonStyle;
    };
    paginationBetween: {
        buttonSkip: ButtonStyle;
        buttonDone: ButtonStyle;
        buttonPrevious: ButtonStyle;
        buttonNext: ButtonStyle;
    };
}

export const defaultWelcomeScreenStyle: IntroScreenStyle = {
    fullscreenContainer: {
        ...absoluteFillObject
    },
    popupContainer: {
        ...absoluteFillObject,
        paddingHorizontal: 50,
        paddingVertical: 150,
        backgroundColor: "rgba(0, 0, 0, 0.5)"
    },
    paginationContainer: {
        position: "absolute",
        bottom: isIphoneWithNotch ? 22 : 0,
        left: 0,
        right: 0,
        justifyContent: "space-between",
        alignItems: "center"
    },
    paginationText: {
        fontSize: 12,
        color: "black"
    },
    paginationAbove: {
        buttonsContainer: {
            flex: 1,
            flexDirection: isAndroidRTL ? "row-reverse" : "row",
            justifyContent: "center",
            marginTop: 16
        },
        buttonSkip: {
            container: {
                ...defaultButtonBelowContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        },
        buttonDone: {
            container: {
                ...defaultButtonBelowContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        },
        buttonPrevious: {
            container: {
                ...defaultButtonBelowContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        },
        buttonNext: {
            container: {
                ...defaultButtonBelowContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        }
    },
    paginationBetween: {
        buttonSkip: {
            container: {
                ...defaultButtonBetweenContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        },
        buttonDone: {
            container: {
                ...defaultButtonBetweenContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        },
        buttonPrevious: {
            container: {
                ...defaultButtonBetweenContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        },
        buttonNext: {
            container: {
                ...defaultButtonBetweenContainer
            },
            caption: {
                ...defaultButtonCaption,
                paddingHorizontal: 5
            },
            icon: {
                color: "black"
            }
        }
    },
    dotStyle: {
        backgroundColor: "rgba(0, 0, 0, .2)"
    },
    activeDotStyle: {
        backgroundColor: "rgba(255, 255, 255, .9)"
    }
};
