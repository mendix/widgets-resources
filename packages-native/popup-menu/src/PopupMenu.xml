<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.popupmenu.PopupMenu" pluginWidget="true" offlineCapable="true"
        supportedPlatform="Native"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../node_modules/mendix/custom_widget.xsd">
    <name>Popup Menu</name>
    <description>My widget description</description>
    <icon/>
    <properties>
        <propertyGroup caption="General">
            <property key="menuTriggerer" type="widgets" required="false">
                <caption>Clickable area</caption>
                <description>Wrapped with touchable area. Responsible for toggling the popupMenu</description>
            </property>
            <property key="renderMode" type="enumeration" defaultValue="basic">
                <caption>Render</caption>
                <description>Basic: creates the menu based on the configured options. Custom: free to model menu.</description>
                <enumerationValues>
                    <enumerationValue key="basic">Basic</enumerationValue>
                    <enumerationValue key="custom">Custom</enumerationValue>
                </enumerationValues>
            </property>
            <property key="basicItems" type="object" isList="true" required="false">
                <caption>Items</caption>
                <description />
                <properties>
                    <property key="itemType" type="enumeration" defaultValue="item">
                        <caption>Item type</caption>
                        <category>General</category>
                        <description />
                        <enumerationValues>
                            <enumerationValue key="item">Item</enumerationValue>
                            <enumerationValue key="divider">Divider</enumerationValue>
                        </enumerationValues>
                    </property>
                    <property key="caption" type="string" required="false">
                        <caption>Caption</caption>
                        <category>General</category>
                        <description />
                    </property>
                    <property key="action" type="action" required="false">
                        <caption>Action</caption>
                        <category>General</category>
                        <description />
                    </property>
                    <property key="styleClass" type="enumeration" defaultValue="default">
                        <caption>Style</caption>
                        <category>General</category>
                        <description />
                        <enumerationValues>
                            <enumerationValue key="default">Default</enumerationValue>
                            <enumerationValue key="primary">Primary</enumerationValue>
                            <enumerationValue key="danger">Danger</enumerationValue>
                        </enumerationValues>
                    </property>
                </properties>
            </property>
            <property key="complexItems" type="object" isList="true" required="false">
                <caption>Items Complex</caption>
                <description />
                <properties>
                    <property key="content" type="widgets">
                        <caption>Content</caption>
                        <category>General</category>
                        <description />
                    </property>
                    <property key="action" type="action" required="false">
                        <caption>Action</caption>
                        <category>General</category>
                        <description />
                    </property>
                </properties>
            </property>
        </propertyGroup>

    </properties>
</widget>
