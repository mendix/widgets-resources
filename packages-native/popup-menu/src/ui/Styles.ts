import { TextStyle, TouchableHighlightProps, ViewStyle } from "react-native";

interface Touchable extends TouchableHighlightProps, ViewStyle {
    container?: ViewStyle;
}

export interface PopupMenuStyle {
    dividerColor?: string;
    menuItem?: PopupItemStyle;
    touchable?: Touchable;
}

/**
 *
 * - `head` - The line is displayed so that the end fits in the container and the missing text
 * at the beginning of the line is indicated by an ellipsis glyph. e.g., "...wxyz"
 * - `middle` - The line is displayed so that the beginning and end fit in the container and the
 * missing text in the middle is indicated by an ellipsis glyph. "ab...yz"
 * - `tail` - The line is displayed so that the beginning fits in the container and the
 * missing text at the end of the line is indicated by an ellipsis glyph. e.g., "abcd..."
 * - `clip` - Lines are not drawn past the edge of the text container.
 *
 * The default is `tail`.
 *
 * `numberOfLines` must be set in conjunction with this prop.
 *
 * > `clip` is working only for iOS
 */
enum EllipsizeMode {
    "head",
    "middle",
    "tail",
    "clip"
}

interface PopupItemStyle {
    // Only for ios
    underlayColor?: string;
    // Item text style
    textStyle?: TextStyle;
    // View surrounding the menu item excluding touchable
    ellipsizeMode?: EllipsizeMode;
    complexItemContainer?: ViewStyle;
    basicItemContainer?: ViewStyle;
}

export const defaultPopupMenuStyles: PopupMenuStyle = {
    dividerColor: "green",
    menuItem: {
        underlayColor: "#e0e0e0",
        textStyle: {
            color: "red"
        },
        complexItemContainer: {
            backgroundColor: "yellow",
            height: 48,
            justifyContent: "center",
            maxWidth: 248,
            minWidth: 124
        }
    },
    touchable: {
        underlayColor: "#e0e0e0"
    }
};

// TODO: request default,primary,danger
