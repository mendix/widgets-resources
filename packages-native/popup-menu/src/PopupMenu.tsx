import { ComponentType, createElement, ReactElement, useCallback, useMemo, useRef } from "react";
import { PopupMenuProps } from "../typings/PopupMenuProps";
import { defaultPopupMenuStyles, PopupMenuStyle } from "./ui/Styles";
import { executeAction } from "@widgets-resources/piw-utils";
import {
    Platform,
    TouchableHighlight,
    TouchableHighlightProps,
    TouchableNativeFeedback,
    TouchableNativeFeedbackProps,
    View
} from "react-native";
import { ActionValue } from "mendix";
import Menu, { MenuDivider, MenuItem } from "react-native-material-menu";
import deepmerge from "deepmerge";

export function PopupMenu(props: PopupMenuProps<PopupMenuStyle>): ReactElement {
    const styles = useMemo(
        () => deepmerge.all<PopupMenuStyle>([defaultPopupMenuStyles, ...props.style]),
        [props.style]
    );

    const menuRef = useRef<any>(null);
    const showMenu = useCallback(() => {
        menuRef.current?.show();
    }, [menuRef]);
    const handlePress = useCallback(
        (action?: ActionValue) => {
            menuRef.current?.hide();
            executeAction(action);
        },
        [menuRef]
    );

    const Touchable: ComponentType<TouchableNativeFeedbackProps | TouchableHighlightProps> =
        Platform.OS === "android" ? TouchableNativeFeedback : TouchableHighlight;

    let menuOptions: ReactElement[];
    if (props.renderMode === "basic") {
        menuOptions = props.basicItems.map((item, index) =>
            item.itemType === "divider" ? (
                <MenuDivider key={index} color={styles.basicItem?.dividerColor} />
            ) : (
                <MenuItem
                    key={index}
                    onPress={() => handlePress(item.action)}
                    underlayColor={styles.basicItem?.underlayColor}
                    textStyle={styles.basicItem?.textStyle}
                    ellipsizeMode={styles.basicItem?.ellipsizeMode as any}
                    style={styles.basicItem?.container as any}
                >
                    {item.caption}
                </MenuItem>
            )
        );
    } else {
        menuOptions = props.complexItems.map((item, index) => (
            <Touchable
                key={index}
                onPress={() => handlePress(item.action)}
                underlayColor={styles.complexItem?.underlayColor}
                {...(Platform.OS === "android"
                    ? {
                          background: TouchableNativeFeedback.SelectableBackground()
                      }
                    : {})}
            >
                <View style={styles.complexItem?.container}>{item.content}</View>
            </Touchable>
        ));
    }

    return (
        <Menu
            ref={menuRef}
            button={
                <Touchable onPress={showMenu} underlayColor={styles.buttonUnderlayColor}>
                    <View style={styles.buttonContainer}>{props.menuTriggerer}</View>
                </Touchable>
            }
        >
            {menuOptions}
        </Menu>
    );
}
