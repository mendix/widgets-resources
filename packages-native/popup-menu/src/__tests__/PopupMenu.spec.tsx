import { PopupMenuProps } from "../../typings/PopupMenuProps";
import { PopupMenuStyle } from "../ui/Styles";
import { Text, TouchableHighlight } from "react-native";
import { createElement } from "react";
import { actionValue } from "@native-mobile-resources/util-widgets";
import { fireEvent, render } from "react-native-testing-library";
import { PopupMenu } from "../PopupMenu";
import { MenuItem } from "react-native-material-menu";

let dummyActionValue: any;
let defaultProps: PopupMenuProps<PopupMenuStyle>;

describe("Popup menu", () => {
    beforeEach(() => {
        dummyActionValue = actionValue();
        defaultProps = {
            renderMode: "basic",
            name: "popup-menu",
            style: [],
            menuTriggerer: undefined,
            basicItems: [
                { itemType: "item", action: dummyActionValue, caption: "yolo", styleClass: "default" },
                { itemType: "divider", styleClass: "default" }
            ],
            complexItems: [{ content: <Text>Yolo</Text>, action: dummyActionValue }]
        };
    });
    it("renders menu triggerer", () => {
        const menuTriggerer = <Text>Menu Triggerer</Text>;
        const component = render(<PopupMenu {...defaultProps} menuTriggerer={menuTriggerer} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    describe("with basic items", () => {
        it("renders", () => {
            const component = render(<PopupMenu {...defaultProps} />);

            expect(component.toJSON()).toMatchSnapshot();
        });

        it("triggers action", () => {
            const component = render(<PopupMenu {...defaultProps} />);
            fireEvent.press(component.getByType(MenuItem));
            expect(dummyActionValue.execute).toHaveBeenCalled();
        });

        it("renders with custom styles", () => {
            const customStyle = [
                {
                    menuItem: {
                        underlayColor: "green",
                        textStyle: {
                            color: "green"
                        }
                    }
                }
            ];
            const component = render(<PopupMenu {...defaultProps} style={customStyle} />);

            expect(component.toJSON()).toMatchSnapshot();
        });
    });

    describe("with complex items", () => {
        beforeEach(() => {
            defaultProps.renderMode = "custom";
        });

        it("renders", () => {
            const component = render(<PopupMenu {...defaultProps} />);

            expect(component.toJSON()).toMatchSnapshot();
        });

        it("triggers action", () => {
            const component = render(<PopupMenu {...defaultProps} />);
            fireEvent.press(component.getAllByType(TouchableHighlight).pop()!);
            expect(dummyActionValue.execute).toHaveBeenCalled();
        });

        it("renders with custom styles", () => {
            const customStyle = [
                {
                    menuItem: {
                        complexItemContainer: {
                            backgroundColor: "yellow"
                        }
                    }
                }
            ];
            const component = render(<PopupMenu {...defaultProps} style={customStyle} />);

            expect(component.toJSON()).toMatchSnapshot();
        });
    });
});
