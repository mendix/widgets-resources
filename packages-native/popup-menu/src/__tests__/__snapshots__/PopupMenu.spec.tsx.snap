// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup menu renders menu triggerer 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        style={null}
      >
        <Text>
          Menu Triggerer
        </Text>
      </View>
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 4,
            "left": 8,
            "opacity": 0,
            "position": "absolute",
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.14,
            "shadowRadius": 2,
            "top": 8,
            "transform": Array [],
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "height": 48,
                    "justifyContent": "center",
                    "maxWidth": 248,
                    "minWidth": 124,
                  },
                  undefined,
                ]
              }
            >
              <Text
                ellipsizeMode="clip"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "fontWeight": "400",
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                    },
                    false,
                    Object {
                      "color": "red",
                    },
                  ]
                }
              >
                yolo
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                },
                Object {
                  "borderBottomColor": "green",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </Modal>
</View>
`;

exports[`Popup menu with basic items renders 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        style={null}
      />
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 4,
            "left": 8,
            "opacity": 0,
            "position": "absolute",
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.14,
            "shadowRadius": 2,
            "top": 8,
            "transform": Array [],
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "height": 48,
                    "justifyContent": "center",
                    "maxWidth": 248,
                    "minWidth": 124,
                  },
                  undefined,
                ]
              }
            >
              <Text
                ellipsizeMode="clip"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "fontWeight": "400",
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                    },
                    false,
                    Object {
                      "color": "red",
                    },
                  ]
                }
              >
                yolo
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                },
                Object {
                  "borderBottomColor": "green",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </Modal>
</View>
`;

exports[`Popup menu with basic items renders with custom styles 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        style={null}
      />
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 4,
            "left": 8,
            "opacity": 0,
            "position": "absolute",
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.14,
            "shadowRadius": 2,
            "top": 8,
            "transform": Array [],
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Array [
                  Object {
                    "height": 48,
                    "justifyContent": "center",
                    "maxWidth": 248,
                    "minWidth": 124,
                  },
                  undefined,
                ]
              }
            >
              <Text
                ellipsizeMode="clip"
                numberOfLines={1}
                style={
                  Array [
                    Object {
                      "fontSize": 14,
                      "fontWeight": "400",
                      "paddingHorizontal": 16,
                      "textAlign": "left",
                    },
                    false,
                    Object {
                      "color": "green",
                    },
                  ]
                }
              >
                yolo
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "borderBottomWidth": 0.5,
                  "flex": 1,
                },
                Object {
                  "borderBottomColor": "green",
                },
              ]
            }
          />
        </View>
      </View>
    </View>
  </Modal>
</View>
`;

exports[`Popup menu with complex items renders 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        style={null}
      />
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 4,
            "left": 8,
            "opacity": 0,
            "position": "absolute",
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.14,
            "shadowRadius": 2,
            "top": 8,
            "transform": Array [],
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "backgroundColor": "yellow",
                  "height": 48,
                  "justifyContent": "center",
                  "maxWidth": 248,
                  "minWidth": 124,
                }
              }
            >
              <Text>
                Yolo
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;

exports[`Popup menu with complex items renders with custom styles 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={null}
    >
      <View
        style={null}
      />
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      accessible={false}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "backgroundColor": "white",
            "borderRadius": 4,
            "left": 8,
            "opacity": 0,
            "position": "absolute",
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.14,
            "shadowRadius": 2,
            "top": 8,
            "transform": Array [],
          }
        }
      >
        <View
          style={
            Object {
              "overflow": "hidden",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "backgroundColor": "yellow",
                  "height": 48,
                  "justifyContent": "center",
                  "maxWidth": 248,
                  "minWidth": 124,
                }
              }
            >
              <Text>
                Yolo
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
